### DISCUSSION TITLE
Add Langchain Support to Prompt Management

### DISCUSSION THREAD

**Discussion by guidev** [langfuse/langfuse#2237] [Ideas]
### Describe the feature or potential improvement

The current documentation notes that prompt management cannot be linked to traces when using Langchain. It would be highly beneficial for users like myself to have this functionality supported.

https://langfuse.com/docs/prompts/get-started

**Comment by marcklingen** [comment]
agree! we are considering options for this. If you have a great idea or are open to contribute to this, happy to discuss/support

**Comment by dewert** [comment]
Currently, you can associate a prompt with a trace in a janky way by doing the following (untested generic example):

```python
@observe()
def do_stuff(message):
    langfuse = Langfuse()
    model = ChatOpenAI()

    langfuse_prompt = _build_prompt()
    prompt = ChatPromptTemplate.from_messages(langfuse_prompt.get_langchain_prompt())
    prompt.append(MessagesPlaceholder("messages"))
    chain = prompt | model
    callback_handler = langfuse_context.get_current_langchain_handler()

    return chain.run(
        {"messages": [HumanMessage(content=message)]},
        config={"callbacks": [callback_handler]},
    )

@observe(as_type="generation")
def _build_prompt(langfuse):
    langfuse_prompt = langfuse.get_prompt("my-chat-prompt")
    langfuse_context.update_current_observation(prompt=langfuse_prompt)

    return langfuse_prompt
```
The problem is of course that `_build_prompt` shows up as a generation, whereas the prompt should be associated with the OpenAI call.

It would be fantastic if the prompt was just picked up as part of the chain, but I don't know if that's feasible. I'd be happy to manage it manually - say if the callback handler was configured with a prompt that it associated with any OpenAI generations it captured.

**Reply by marcklingen** [reply]
> say if the callback handler was configured with a prompt that it associated with any OpenAI generations it captured.

this could be the workaround solution here, although it is a bit limited as it cannot handle multiple different prompts within a langchain application -- which will mostly be the case

**Reply by dewert** [reply]
Thanks, yes, that's a very good point! I'm new to this, so I've only done some simple tests.

In that case, I could still see some use if there was at least a prompt -> trace link, even if it wasn't immediately obvious which prompts were associated with which generations, but that might not line up well with the eventual goal.

**Reply by ghost** [reply]
Since the TS library doesn't have the `get_current_langchain_handler` it seems like we should find another way to associate a trace with a prompt. I'll take a crack at this and report back.

**Reply by marcklingen** [reply]
the TS library accepts a `root` trace in the CallbackHandler constructor, see example here: https://langfuse.com/docs/integrations/langchain/tracing#interoperability

**Reply by ghost** [reply]
Right @marcklingen , thanks!

I was able to construct a similar workaround as @dewert by doing the following.
```js
// Initialize langfuse trace and callback handler
const langfuse = new Langfuse({...config});
const trace = langfuse.trace({...traceConfig});
const callback = new CallbackHandler({
  root: trace,
  updateRoot: false,
});

// Observe prompt generation in trace
const generation = trace.generation({
  name: "Prompt",
});
const prompt = await langfuse.getPrompt(promptName);
generation.end({
  prompt: prompt,
  output: prompt,
});

// Now use the callback in langchain as normal
const result = await llm.invoke(messages, {
  callbacks: [callback],
})
```

This will result in a trace as so, where the `Prompt` generation is linked to the langfuse prompt in question.
![image](https://github.com/user-attachments/assets/266e3a40-605a-4815-b7d0-8260f6603f70)

Of course, this workaround has the same limitations as the Python one, but for simple single-prompt chains this seems to work well enough.

**Reply by marcklingen** [reply]
thanks for sharing the code example! agree, this is a workaround but it works to link the version, however, you will not be able to see metrics by prompt version as this does not link the generation itself

**Reply by ghost** [reply]
That's a fair remark @marcklingen.
Am I wrong, or does @dewert's original example have the same issue.
If you see any way of resolving this, please let me know. I don't mind spending some more time testing things out.

**Reply by ghost** [reply]
The only (less or more clean) way I found out to link the prompt straight to the generation is by subclassing and overwriting the callback handler for now.

```js
type PromptInfo = {
  promptName: string;
  promptVersion: number;
};

class CallbackHandlerWithPromptInfo extends CallbackHandler {
  promptInfo: PromptInfo | null = null;

  setPromptInfo(promptInfo: PromptInfo) {
    this.promptInfo = promptInfo;
  }

  addPromptInfoToGeneration(runId: string) {
    if (this.promptInfo) {
      this.langfuse._updateGeneration({
        id: runId,
        traceId: this.traceId,
        ...this.promptInfo,
      });
    }
  }

  async handleLLMEnd(
    output: LLMResult,
    runId: string,
    parentRunId?: string | undefined,
  ): Promise<void> {
    // Log LLM Result as normal
    await super.handleLLMEnd(output, runId, parentRunId);
    // Add prompt info to generation
    this.addPromptInfoToGeneration(runId);
  }

  async handleLLMError(
    err: any,
    runId: string,
    parentRunId?: string | undefined,
  ): Promise<void> {
    // Log LLM Error as normal
    await super.handleLLMError(err, runId, parentRunId);
    // Add prompt info to generation
    this.addPromptInfoToGeneration(runId);
  }
}
```

This way, it can be used as such:

```js
// Create callback
const callback = new CallbackHandlerWithPromptInfo({...options});

// Fetch prompt
const prompt = await callback.langfuse.getPrompt(id)

// Set prompt info on callback
const promptInfo = {
  promptName: prompt.name,
  promptVersion: prompt.version,
};
callback.setPromptInfo(promptInfo);

// Use callback as normal
const result = await llm.invoke(messages, {
  callbacks: [callback],
})
```

Note that for every new prompt, you have to set the new prompt info in your callback.

**Comment by yusuke-intern** [comment]
Upvote!

**Comment by AyushSonuu** [comment]
iam happy to contribute to this

**Comment by hassiebp** [comment]
Linking Langfuse prompts to Langchain executions has been one of the most requested features recently, and we have just released support for it 🫡 🥳

For more details, please check out:

- [Docs on how to integrate, see tabs for Langchain in examples](https://langfuse.com/docs/prompts/get-started#link-with-langfuse-tracing-optional)
- [Python SDK release](https://github.com/langfuse/langfuse-python/releases/tag/v2.49.0)
- [JS SDK release](https://github.com/langfuse/langfuse-js/releases/tag/v3.25.0)

**Reply by marcklingen** [reply]
Big one, thanks @hassiebp for the update!
